text/0000-value-with-callback-return.md

```diff
@@ -0,0 +1,336 @@
+}
+
+// Attaches a callback back to `A`. (R2 is modified, R3 is created)
+ActionReceipt {
+    id: "R2",
+    receiver_id: "B",
+    predecessor_id: "A",
+    input_data_ids: [],
+    output_data_receivers: [
+        DataReceiver {receiver_id: "A", data: "D1"}
+    ]
+}
+ActionReceipt {
+    id: "R3",
+    receiver_id: "A",
+    predecessor_id: "A",
+    input_data_ids: ["D1"],
+    output_data_receivers: []
+}
+
+// Returns callback `A`. (Doesn't change anything, cause R1 doesn't have output)
+
+
+//////// Executing R2
+
+// `B` calls `C` and `D`. (R4 and R5 are created)
+ActionReceipt {
+    id: "R4",
+    receiver_id: "C",
+    predecessor_id: "B",
+    input_data_ids: [],
+    output_data_receivers: [],
+}
+ActionReceipt {
+    id: "R5",
+    receiver_id: "D",
+    predecessor_id: "B",
+    input_data_ids: [],
+    output_data_receivers: [],
+}
+
+// Attaches a callback from `C` to `D`. (R4 and R5 are modified)
+ActionReceipt {
+    id: "R4",
+    receiver_id: "C",
+    predecessor_id: "B",
+    input_data_ids: [],
+    output_data_receivers: [
+        DataReceiver {receiver_id: "D", data: "D2"},
+    ],
+}
+ActionReceipt {
+    id: "R5",
+    receiver_id: "D",
+    predecessor_id: "B",
+    input_data_ids: ["D2"],
+    output_data_receivers: [],
+}
+// And returns promise `C`. (R4 is modified)
+ActionReceipt {
+    id: "R4",
+    receiver_id: "C",
+    predecessor_id: "B",
+    input_data_ids: [],
+    output_data_receivers: [
+        DataReceiver {receiver_id: "A", data: "D1"},
+        DataReceiver {receiver_id: "D", data: "D2"},
+     ],
+}
+```
+
+So now when `R4` is executed, it will send 2 data receipts.
+
+Let's now discuss how to implement the proposed change.
+
+## Proposed change
+
+### Back to example
+
+In the example with `fun` tokens, we need to attach a promise on the caller. But let's look at the example of `R4` receipt instead:
+```rust
+ActionReceipt {
+    id: "R4",
+    receiver_id: "C",
+    predecessor_id: "B",
+    input_data_ids: [],
+    output_data_receivers: [
+        DataReceiver {receiver_id: "A", data: "D1"},
+        DataReceiver {receiver_id: "D", data: "D2"},
+     ],
+}
+```
+
+There caller (`predecessor_id`) is `B`, but the output data receivers are `A` and `D`.
+Execution at `C` can't influence `B` or attach anything to `B`, because execution of `B` has already completed or has indirect dependency.
+Instead `C` can only influence both `A` and `D`. 
+
+Now lets look at `fun` token receipt example:
+```rust
+ActionReceipt {
+    id: "R6",
+    receiver_id: "fun",
+    predecessor_id: "dex",
+    input_data_ids: [],
+    output_data_receivers: [
+        DataReceiver {receiver_id: "dex", data: "D3"},
+     ],
+}
+```
+
+In this case `dex` is both the caller and the output data receiver.
+So for the `lock` example `fun` might be able to attach something towards `dex` through the generated output data.
+
+### Changes
+
+- Add a new runtime API method `value_with_callback_return` (the name is debatable).
```

Comment:
I am still not sure why we need to modify the runtime API. Why cannot we resolve it by adding `acquires_lock`, `releases_lock`, `requires_lock` to decorate the internals of the token contract, and then on the exchange side we can have something like `callback_args( (lock0, data0), (lock1, data1))` decorator so when it calls `nai` it can use `nai::transfer(lock0, ...)`?

Reply:
We need to have automatic unlock on `nai` in case `on_locks` fails